### PR TITLE
Add support for keccak 224, 256, 384 and 512.

### PR DESCRIPTION
**Pull request by setine**
Wondering if this is something you'd like to have upstream, given the existing issue below.  Happy to work further on this if there's interest.

Fixes issue openssl#13033

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
The slight problem is that the original Keccak is not standardized and thus it would go against rules for inclusion of new algorithm support. I suppose OTC or OMC would have to give an exception.

**Comment by t8m** [comment]
You will also have to submit a CLA otherwise we cannot merge this.
https://www.openssl.org/policies/cla.html

**Comment by setine** [comment]
Thanks for taking a look. Makes sense, happy to go through the CLA process if we can get an exception for this. I'll wait for an update here.

**Comment by t8m** [comment]
The CI failures are relevant.

**Comment by setine** [comment]
I suppose it doesn't make sense to add these to the FIPS provider, so I've restricted the tests to the default provider.

**Comment by t8m** [comment]
The cross compilation CI failure is still relevant.

**Comment by paulidale** [comment]
You're correct, I don't think these can be validated by NIST, so they shouldn't be in the FIPS provider.

**Comment by setine** [comment]
This resolved the S390X build issues locally for me, though I didn't manage to run the Qemu tests. Looking at [another run](https://github.com/openssl/openssl/runs/3611268499?check_suite_focus=true), it looks like they're not being run in CI either.

**Comment by paulidale** [comment]
To get the qemu tests to run, you need to commit a temporary change to `.github/workflows/cross-compiles.yml` removing the three instances of `github.event_name == 'push' && `.  These prevent the tests from executing for pull requests.

File: test/recipes/30-test_evp_data/evpmd_sha.txt [hunk 1]
```diff
@@ -281,6 +281,99 @@ Digest = SHAKE256
 Input = 8d8001e2c096f1b88e7c9224a086efd4797fbf74a8033a2d422a2b6b8f6747e4
 Output = 2e975f6a8a14f0704d51b13667d8195c219f71e6345696c49fa4b9d08e9225d3d39393425152c97e71dd24601c11abcfa0f12f53c680bd3ae757b8134a9c10d429615869217fdd5885c4db174985703a6d6de94a667eac3023443a8337ae1bc601b76d7d38ec3c34463105f0d3949d78e562a039e4469548b609395de5a4fd43c46ca9fd6ee29ada5efc07d84d553249450dab4a49c483ded250c9338f85cd937ae66bb436f3b4026e859fda1ca571432f3bfc09e7c03ca4d183b741111ca0483d0edabc03feb23b17ee48e844ba2408d9dcfd0139d2e8c7310125aee801c61ab7900d1efc47c078281766f361c5e6111346235e1dc38325666c
 
+
```
Comment (slontis) [inline test/recipes/30-test_evp_data/evpmd_sha.txt:284]: Please add a reference to where the test vectors came from in a comment here

**Comment by setine** [comment]
> To get the qemu tests to run, you need to commit a temporary change to `.github/workflows/cross-compiles.yml` removing the three instances of `github.event_name == 'push' && `. These prevent the tests from executing for pull requests.

Thanks, added a temporary commit for that. Out of curiosity, is the qemu setup available somewhere? I'd be interested in running it locally, as well.

File: test/recipes/30-test_evp_data/evpmd_sha.txt (hunk 1 above)
Comment (setine) [reply test/recipes/30-test_evp_data/evpmd_sha.txt:284]: Good point, these were from [1], so I've replaced them now with canonical[2] ones and documented that.

[1] https://emn178.github.io/online-tools/keccak_256.html
[2] https://keccak.team/archives.html

**Comment by paulidale** [comment]
Yep, all in the cross-compiles.  It's a bit hairy in places but it hangs together.

There are additional flavours/compilers/platforms that could be added.  I think that enough of the major (and not so major) processors are covered.

**Comment by paulidale** [comment]
OTC: this doesn't conform to our policy about being a standard before inclusion. In this case, an exception vote has taken place and this can be included. The reason being that it is in use, we have the implementation and that it just needs to be exposed.

**Comment by setine** [comment]
That's great, thanks for the update! I've added what I believe is the appropriate documentation for this, please let me know if I missed anything.
(Signed CLA sent by email, too.)

**Review by paulidale — APPROVED** [review]

**Review by t-j-h — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for this contribution.

**Comment by setine** [comment]
Happy to, thanks for the reviews!

**Pull request closed without merging**